### PR TITLE
(perf-tweak-4) Overmap related tweaks for big improvements.

### PR DESCRIPTION
**Pull request by bkentel**
A relatively small set of overmap related performance tweaks, and a tiny bit of cleanup in the related code.
These changes make a huge impact on the performance of the map screen (in a debug build, 3~5x faster for me), and a minor improvement to performance in general.

**catacurse.h**
- use compiler generated assignment.
- use in class initializers.
- tweak equality comparison order to look at the string last.
- use `std::move()`

**omdata.h**
- **Performance** return a const ref, not a copy.

**overmap.cpp**
- factor out note related parsing into `get_note_display_info()`.
- **Performance** Use a small LRU cache to avoid lookups in the `unordered_map`. Most terrain is "clumpy", so even just a tiny cache improved performance significantly.
- move the check for npcs etc into the actual if statements.
- small cleanup of `corner_text` related code.

**overmapbuffer.cpp**
- **Performance** use `last_requested_overmap` in `overmapbuffer::get()` too to avoid costly lookups.
- **New** `overmapbuffer::has_horde()` to mirror the related `has_*` functions.
- **Performance** implement `has_*` in a less costly manner to avoid all kinds of redundant work.
- **Performance** `overmapbuffer::get_overmaps_near` almost never returns more than a tiny handful of items. Eliminated the `std::set` and just sorted a vector.

**sdltiles.cpp**
- Change `OutputChar()` to take it's first parameter by value; it's always moved from internally anyway.
- **Performance** Change `glyph_cache_map` to use a custom key and value. `key_t` compares by color first, not string. `cached_t` now caches the display width instead of calculating it redundantly on every render.
- **Performance** keys are constructed more carefully, and the insert if not found behavior properly uses the `hint` parameter now.

See #11180

### CODE REVIEW

File: src/overmap.cpp [hunk 1]
```diff
@@ -1348,139 +1385,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
-                        }
-                        ter_sym = otermap[cur_ter].sym;
+                        // cache the new value
+                        info = &it->second;
+                        cache[cache_next] = std::make_pair(cur_ter, info);
+                        cache_next = (cache_next + 1) % cache_size;
                     }
                 }
+                // Ok, we found something
+                if (info) {
```
Comment (DavidKeaton) [inline src/overmap.cpp]: minor critique, since caching is introduced, bring the if statement for it closer so it's less of a jump? I am sure they are both near jumps, but perhaps it can be optimized for jump prediction?

Comment (bkentel) [inline src/overmap.cpp]: Hmm, I wonder if I can do that cleanly without introducing some code duplication -- not sure the branch prediction would make a huge difference here considering all the other surrounding code though.

**Comment by DavidKeaton** [comment]
Fair enough, I'm a crazy assembler guy. :-)

File: src/overmap.cpp [hunk 2]
```diff
@@ -1348,139 +1385,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
-        for (int j = 0; j < om_map_height; j++) {
-            const int omx = cursx + i - (om_map_width / 2);
-            const int omy = cursy + j - (om_map_height / 2);
+    // A small LRU cache: most oter_id's occur in clumps like forests of swamps.
+    // This cache helps avoid much more costly lookups in the full hashmap.
+    constexpr size_t cache_size = 8;
+    std::array<std::pair<oter_id, oter_t const*>, cache_size> cache {};
+    size_t cache_next = 0;
+
+    auto const offset_x = cursx - (om_map_width  / 2);
```
Comment (kevingranade) [inline src/overmap.cpp]: You had indicated you were going to stop using auto for primitive types, do so.

Comment (bkentel) [inline src/overmap.cpp]: I'll fix that up; rolls off the fingers out of habit sometimes, sorry. Other than this, is there anything else you noticed that might hold up this PR?

Comment (kevingranade) [inline src/overmap.cpp]: Just the size.

Comment (bkentel) [inline src/overmap.cpp]: The `cache_size` variable? I stuck that in there because of the rather larger separation between here and where `cache_size` is used again to calculate the next index. I could remove it, but if the size changes you'd have to be careful to update both locations properly.

**Comment by bkentel** [comment]
Since it seems you guys have been purging a lot of `auto` I'll go through this and flip some of the ones you might object to.

**Comment by DavidKeaton** [comment]
I have begun seeing the use of auto, but for class or structs, it can be a bit irritating. :-)

File: src/overmap.cpp (hunk 2 above)
Comment (kevingranade) [inline src/overmap.cpp]: No, the size of the PR, the bigger it is the longer it's going to take to get merged, sometimes exponentially.

File: src/overmap.cpp [hunk 3]
```diff
@@ -3713,13 +3660,14 @@ const unsigned &oter_id::operator=(const int &i)
     return _val;
 }
 // ter(...) = "rock"
-oter_id::operator std::string() const
+namespace { std::string const debug_dummy_string; }
```
Comment (BevapDin) [inline src/overmap.cpp]: Why this and not just a static variable inside the function?

Comment (bkentel) [inline src/overmap.cpp]: It could be, but since c++11, function local statics have had their initialization and access rules updated to play well in cases where race conditions can exist. This causes some extra overhead, and since this was ostensibly a tweak made for performance (and I don't see CATA going multithreaded anytime soon...)

[Assembly listing](http://pastebin.com/MWfLUEe3)

On the other hand, the impact is probably pretty minor and I could move this inside the function.

Comment (BevapDin) [inline src/overmap.cpp]: You've got an interesting point there, I wasn't aware of that.

But the static variable could be directly after the debug message, which requires user input and is therefor not performance relevant code. And it should not happen anyway, so the static variable (inside the if block) should never be initialized.

Comment (bkentel) [inline src/overmap.cpp]: A good point and probably the best solution here. I'll make the change.

File: src/overmap.cpp [hunk 4]
```diff
@@ -1348,139 +1385,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
         }
     }
 
-    for (int i = 0; i < om_map_width; i++) {
-        for (int j = 0; j < om_map_height; j++) {
-            const int omx = cursx + i - (om_map_width / 2);
-            const int omy = cursy + j - (om_map_height / 2);
+    // A small LRU cache: most oter_id's occur in clumps like forests of swamps.
```
Comment (kevingranade) [inline src/overmap.cpp]: How'd this do in benchmarking?
You did test performance with and without it, right?

File: src/overmap.cpp [hunk 5]
```diff
@@ -1561,81 +1554,35 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
-        if( npc->marked_for_death ) {
-            continue;
-        }
-        corner_text.push_back( npc->name );
-    }
-    const auto vehicles = overmap_buffer.get_vehicle(cursx, cursy, z);
-    for( const auto &v : vehicles ) {
-        corner_text.push_back( v.name );
+    for( auto &v : overmap_buffer.get_vehicle(cursx, cursy, z) ) {
+        corner_text.emplace_back( std::move(v.name) );
```
Comment (kevingranade) [inline src/overmap.cpp]: It's not remotely clear that std::move() is safe here, and even if it is now, it's incredibly fragile, just make a copy.
Also, it's just a string, sheesh.

File: src/overmap.cpp [hunk 6]
```diff
@@ -1561,81 +1554,35 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
     }
 
     if( !corner_text.empty() ) {
-        int maxlen = 0;
-        for( const auto &line : corner_text ) {
-            maxlen = std::max( maxlen, utf8_width( line.c_str() ) );
-        }
+        int const maxlen = std::accumulate(begin(corner_text), end(corner_text), int {0},
```
Comment (kevingranade) [inline src/overmap.cpp]: It's unobvious how this works, revert to the original loop.

File: src/overmapbuffer.cpp [hunk 7]
```diff
@@ -464,29 +515,35 @@ std::vector<npc*> overmapbuffer::get_npcs_near_player(int radius)
     return get_npcs_near(plpos.x, plpos.y, plpos.z, radius);
 }
 
-std::vector<overmap *> overmapbuffer::get_overmaps_near( point location, int radius )
+std::vector<overmap*> overmapbuffer::get_overmaps_near( point const location, int const radius )
 {
     // Grab the corners of a square around the target location at distance radius.
     // Convert to overmap coordinates and iterate from the minimum to the maximum.
-    std::set<point> distinct_corners;
```
Comment (kevingranade) [inline src/overmapbuffer.cpp]: What's the rationale for this change? The sort and erase at the end seems quite indirect compared to simply enumerating the overmap coordinates.

Comment (bkentel) [inline src/overmapbuffer.cpp]: I believe this was an early change I made before I found other ways to work even calling this rather heavy function at all. The rational at the time was that, while both the `set` version and the `vector` versions are O(n・logn), especially for small a small n the `set` version doesn't perform particularly well (poor locality of reference, multiple memory allocations, etc).

I would have to profile things again, but this could potentially be changed back to the set version without affecting performance too much if the vector version seems too roundabout. Better yet, there's likely a smarter algorithm for doing this; I merely transformed the code.

File: src/overmap.cpp (hunk 5 above)
Comment (bkentel) [inline src/overmap.cpp]: Fair enough -- it certainly isn't very clear from just the name `get_vehicle` that a copy is returned. Strings are probably short anyway and except for GCC pre 5.0 the small string optimization is used by other compilers anyway.

File: src/overmap.cpp (hunk 4 above)
Comment (bkentel) [inline src/overmap.cpp]: Yes, all of the changes were made piecemeal one by one with measurable results: measure, change, measure. This was one of the bigger improvements if memory serves -- on the order of a 10~15% relative (not global) improvement, but exact numbers I no longer have. In the future perhaps each individual small optimization should be commited separately with a rough estimate of how much it affects performance.

File: src/overmap.cpp (hunk 6 above)
Comment (bkentel) [inline src/overmap.cpp]: It's not really more terse than the original either -- I'll flip it back to a range-for.

**Comment by DavidKeaton** [comment]
loving the refactoring for performance so far btw!

**Comment by bkentel** [comment]
Updated per feedback.

**Comment by kevingranade** [comment]
Trying to catch up with PRs, this one has to be handled separately as a performance change.

**Comment by bkentel** [comment]
No problem -- thanks for letting me know what the hold up is though.

File: src/overmap.cpp [hunk 8]
```diff
@@ -1352,139 +1389,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
+                std::tie(ter_sym, ter_color, std::ignore) =
+                    get_note_display_info(overmap_buffer.note(cur_pos));
             } else if (!see) {
                 // All cases above ignore the seen-status,
                 ter_color = c_dkgray;
-                ter_sym = '#';
+                ter_sym   = '#';
                 // All cases below assume that see is true.
-            } else if (blink && npc_here) {
+            } else if (blink && overmap_buffer.has_npc(omx, omy, z)) {
```
Comment (kevingranade) [inline src/overmap.cpp]: Ok wow, just this should help a ton, in my testing has_npc() was taking 40% of execution time.

**Comment by kevingranade** [comment]
Performance testing is very positive, this is good to go, I'm just waiting to not step on a merge KA101 is doing.

**Comment by KA101** [comment]
Broken 64-bit and Windows builds.  :-(

File: src/overmap.cpp [hunk 9]
```diff
@@ -1352,139 +1389,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
...
     }
 
-    for (int i = 0; i < om_map_width; i++) {
-        for (int j = 0; j < om_map_height; j++) {
-            const int omx = cursx + i - (om_map_width / 2);
-            const int omy = cursy + j - (om_map_height / 2);
+    // A small LRU cache: most oter_id's occur in clumps like forests of swamps.
+    // This cache helps avoid much more costly lookups in the full hashmap.
+    constexpr size_t cache_size = 8; // used below to calculate the next index
+    std::array<std::pair<oter_id, oter_t const*>, cache_size> cache {};
```
Comment (KA101) [inline src/overmap.cpp]: http://stackoverflow.com/questions/14178264/c11-correct-stdarray-initialization

**Comment by bkentel** [comment]
What was the error message @KA101 ? Or was it just a warning? Regardless, I'll put together a PR to fix it.

**Comment by bkentel** [comment]
@KA101 looking at the commit log it looks like you took care of it in the meantime while I was away. Sorry for the inconvenience (doubly so given how you had mentioned that you were a little irked at some other recent build issues).

**Comment by KA101** [comment]
Meh, I inadvertently merged a separate build-breaker last night/this morning too.  So that kinda cancels out the lost sleep and five PRs not-tested.  Just check on that sort of thing next time, OK?
